src/compiler/checker.ts

```diff
@@ -1842,7 +1842,12 @@ module ts {
                     var propName = "" + indexOf(pattern.elements, declaration);
                     var type = isTupleLikeType(parentType) ? getTypeOfPropertyOfType(parentType, propName) : getIndexTypeOfType(parentType, IndexKind.Number);
                     if (!type) {
-                        error(declaration, Diagnostics.Type_0_has_no_property_1, typeToString(parentType), propName);
+                        if (isTupleLikeType(parentType)) {
+                            var length = (<TupleType>parentType).elementTypes.length;
+                            error(declaration, Diagnostics.Tuple_type_0_with_length_1_can_t_be_assigned_to_Tuple_with_length_2, typeToString(parentType), length, pattern.elements.length);
+                        } else {
+                            error(declaration, Diagnostics.Type_0_has_no_property_1, typeToString(parentType), propName);

```

Comment:
This bad message can now only happen if the user removes the indexer from Array, and then writes code like the following:
```ts
var a = [];
var [b, c] = a;
```